File: version4j/src/main/java/com/only4/algorithm/version4j/leetcode/SearchRange.java
package com.only4.algorithm.version4j.leetcode;

/**
 * @author zhenyu.jiang
 */
public class SearchRange {
    public int[] searchRange(int[] nums, int target) {
        int leftBound = findLeftBound(nums, target);
        int rightBound = findRightBound(nums, target);
        return new int[]{leftBound, rightBound};
    }

    private int findLeftBound(int[] nums, int target) {
        int left = 0, right = nums.length - 1;
        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (nums[mid] == target) {
                right = mid - 1;
            } else if (nums[mid] < target) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }

        if (left < 0 || left >= nums.length || nums[left] != target) {
            return -1;
        }
        return left;
    }

    private int findRightBound(int[] nums, int target) {
        int left = 0, right = nums.length - 1;
        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (nums[mid] == target) {
                left = mid + 1;
            } else if (nums[mid] < target) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }

        if (right < 0 || right >= nums.length || nums[right] != target) {
            return -1;
        }
        return right;
    }
}

File: version4k/src/main/kotlin/com/only4/algorithm/version4k/leetcode/searchRange.kt
package com.only4.algorithm.version4k.leetcode

fun searchRange(nums: IntArray, target: Int): IntArray {
    if (nums.isEmpty()) return intArrayOf(-1, -1)

    fun findLeftBound(): Int {
        var left = 0
        var right = nums.size - 1

        while (left <= right) {
            val mid = left + (right - left) / 2
            when {
                nums[mid] == target -> right = mid - 1
                nums[mid] < target -> left = mid + 1
                else -> right = mid - 1
            }
        }

        return if (left >= 0 && left < nums.size && nums[left] == target) left else -1
    }

    fun findRightBound(): Int {
        var left = 0
        var right = nums.size - 1

        while (left <= right) {
            val mid = left + (right - left) / 2
            when {
                nums[mid] == target -> left = mid + 1
                nums[mid] < target -> left = mid + 1
                else -> right = mid - 1
            }
        }

        return if (right >= 0 && right < nums.size && nums[right] == target) right else -1
    }

    return intArrayOf(findLeftBound(), findRightBound())
}
